src/main/java/org/openstreetmap/atlas/geography/atlas/raw/slicing/RawAtlasCountrySlicer.java

```diff
@@ -1,102 +1,311 @@
+
     // The countries to be sliced with
     private final CountryBoundaryMap countryBoundaryMap;
 
     // The boundaries used for slicing
     private final Set<String> countries = new HashSet<>();
 
+    private final List<Shard> loadedShards = new ArrayList<>();
+
+    // Bring in all points -- this is important for proper sectioning later downstream, otherwise we
+    // will bring in geometry definitions with no underlying points
+    private static final Predicate<AtlasEntity> pointPredicate = entity -> entity instanceof Point;
+
+    // Bring in all relations that are tagged as Water or Coastline
+    private static final Predicate<AtlasEntity> relationPredicate = entity -> entity.getType()
+            .equals(ItemType.RELATION)
+            && Validators.isOfType(entity, NaturalTag.class, NaturalTag.WATER,
+                    NaturalTag.COASTLINE);
+
+    // Dynamic expansion filter will be a combination of the water relations and points
+    private static final Predicate<AtlasEntity> dynamicAtlasExpansionFilter = entity -> pointPredicate
+            .test(entity) || relationPredicate.test(entity);
+
     /**
-     * Slices against the given country and boundary map. Note: Will assume the entire world to be
-     * the target MultiPolygon to use.
+     * The default constructor for the old, pre-water relation pipeline-- this method will slice
+     * without dynamic expansion along water relations. Left in for legacy compatibility.
      *
      * @param countries
-     *            The Set of countries to be sliced against
+     *            the list of countries to slice
      * @param countryBoundaryMap
-     *            The {@link CountryBoundaryMap} to use when slicing
+     *            The country boundary map
      */
     public RawAtlasCountrySlicer(final Set<String> countries,
             final CountryBoundaryMap countryBoundaryMap)
     {
         this.countries.addAll(countries);
         this.countryBoundaryMap = countryBoundaryMap;
+        this.sharding = null;
+        this.atlasFetcher = null;
+    }
+
+    /**
+     * Updated constructor for the water relation pipeline-- this method will use the Atlas fetcher
+     * function and sharding tree to dynamically expand on all water relations so they can be sliced
+     * appropriately.
+     *
+     * @param country
+     *            The list of countries to slice. Should really just be one country?
+     * @param countryBoundaryMap
+     *            The country boundary map
+     * @param sharding
+     *            The sharding tree
+     * @param atlasFetcher
+     *            A function return an atlas for a given shard. NOTE: The function expects line
+     *            sliced Atlases. And since one shard can have multiple country sliced Atlas files,
+     *            the fetcher should MultiAtlas these line sliced country Atlas files together
+     *            before returning.
+     */
+    public RawAtlasCountrySlicer(final Set<String> country,
```

Comment:
Can this constructor be combined with the one on line 155? They are doing identical work, one is for one country and another for a set of countries.